lib/elixir/lib/calendar/date/range.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Date.Range do
+  @moduledoc """
+  Defines a range of dates.
+
+  A range of dates represents a discrete number of dates where
+  the first and last values are dates with matching calendars.
+
+  Ranges of dates can be either increasing (`first <= last`) or
+  decreasing (`first > last`). They are also always inclusive.
+
+  A range of dates implements the `Enumerable` protocol, which means
+  functions in the `Enum` module can be used to work with
+  ranges:
+
+      iex> range = Date.range(~D[2001-01-01], ~D[2002-01-01])
+      iex> Enum.count(range)
+      366
+      iex> Enum.member?(range, ~D[2001-02-01])
+      true
+      iex> Enum.reduce(range, 0, fn(_date, acc) -> acc - 1 end)
+      -366
+
+  """
+
+  defstruct [:first, :last, :kind, :first_rata_die, :last_rata_die]
+
+  def new(%{calendar: fc} = first, %{calendar: lc} = last) when fc == lc do
+    %Date.Range{
+      first: first,
+      last: last,
+      first_rata_die: Date.to_rata_die(first),
+      last_rata_die: Date.to_rata_die(last),
+      kind: get_range_kind(first, last),
+    }
+  end
+  def new(_, _) do
+    raise ArgumentError,
+      "Both dates must have matching calendars"
+  end
+
+  defp get_range_kind(first, last) when first > last, do: :desc
+  defp get_range_kind(_first, _last), do: :asc
+
+  defimpl Enumerable, for: Date.Range do
+    def member?(range, date) do
+      date_rata = Date.to_rata_die(date)
+      do_member?(range, date_rata)
+    end
+
+    defp do_member?(%Date.Range{first_rata_die: {frd, _}, last_rata_die: {lrd, _}, kind: :asc}, {drd, _}) do
+      {:ok, drd >= frd && drd <= lrd}
+    end
+
+    defp do_member?(%Date.Range{first_rata_die: {frd, _}, last_rata_die: {lrd, _}, kind: :desc}, {drd, _}) do
+      {:ok, drd <= frd && drd >= lrd}
+    end
+
+    def count(%Date.Range{first_rata_die: {frd, _}, last_rata_die: {lrd, _}}) do
+      {:ok, abs(frd - lrd) + 1}
+    end
+
+    def reduce(%Date.Range{first_rata_die: {frd, _}, last_rata_die: {lrd, rest}, first: %{calendar: c}}, acc, fun) do
+      reduce(frd..lrd, acc, &(fun.(Date.from_rata_die({&1, rest}, c), &2)))
+    end
+
+    def reduce(first..last, acc, fun) do
```

Comment:
It is fine as is. Just remove this clause as mentioned above and we should be fine.